docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,663 @@
+to know how many elements should the `ExpandType` expand into just from the pattern type itself. Fortunately, we still kept a reference to
+the generic type param decl through the `capture` argument in the `ExpandType`. This will allow us to evaluate it into `(int, int, int)` when
+we apply substitution `T=ConcreteTypePack(float, double, void)` to it.
+
+Let's take a look at another more contrived example to understand the substitution process. Assume we have:
+
+```
+interface IFoo
+{
+    associatedtype Assoc;
+};
+struct Foo : IFoo
+{
+    typealias Assoc = int;
+};
+struct Foo2 : IFoo
+{
+    typealias Assoc = float;
+};
+typealias MyPack<each T : IFoo> = expand (each T).Assoc;
+typealias Pack2 = MyPack<Foo, Foo2>;
+```
+
+When evaluating `Pack2`, we will first form a `ConcreteTypePack(Foo, Foo2)` and use it to substitute the `T` parameter in `MyPack`. This will result in `MyPack<ConcreteTypePack(Foo, Foo2)>`.
+Then we continue to resolve this type alias by substituting `expand (each T).Assoc` with `T = ConcreteTypePack(Foo, Foo2)`. 
+
+The expression `expand (each T).Assoc` is translated into
+
+```
+ExpandType(
+    pattern =
+        DeclRefType(
+            LookupDeclRef(
+                EachType(DeclRefType(GenericTypePackParamDecl "T")),
+                IFoo::assoc
+            )
+        ),
+    capture = DeclRefType(GenericTypePackParamDecl "T")
+)
+```
+
+Substituting this with `DeclRefType(T) = ConcreteTypePack(Foo, Foo2)` we will get:
+
+```
+ExpandType(
+    pattern =
+        DeclRefType(
+            LookupDeclRef(
+                EachType(ConcreteTypePack(Foo, Foo2)),
+                IFoo::assoc
+            )
+        ),
+    capture = ConcreteTypePack(Foo, Foo2)
+)
+```
+
+Since the captured type pack in the `ExpandType` is already a concrete type pack, we should be able to turn this `ExpandType` into a
+`ConcreteTypePack`, by substituting `pattern` twice, with `EachType(...)` replaced with a corresponding element in the input `ConcreteTypePack` to form:
+```
+ConcreteTypePack(
+    DeclRefType(
+            LookupDeclRef(
+                Foo,
+                IFoo::assoc
+            )
+        ),
+    DeclRefType(
+            LookupDeclRef(
+                Foo2,
+                IFoo::assoc
+            )
+        )
+)
+```
+
+And by resolving the `LookupDeclRef`, we will get:
+```
+ConcreteTypePack(
+    DeclRefType(StructDecl "int"),
+    DeclRefType(StructDecl "float")
+)
+```
+
+Which is the correct representation for type pack `(int, float)`.
+
+#### Simplification Rules of `Expand` and `Each` Types
+
+By the definition of `expand` and `each`, we have these simplification rules:
+- `expand each T` => `T`
+- `each expand T` => `T`
+
+
+### Type Constraints for Subtype Relationships
+
+We define the sub-type relationship for type packs so that: if `T` is a type pack, then
+`T: IBase` if every type in `T` is a subtype of `IFoo`.
```

Comment:
Could you please reword this sentence a little bit? I don't quite understand what it means.

Reply:
Done.